### PR TITLE
Fix compiling on OpenWrt

### PR DESCRIPTION
**Pull request by PolynomialDivision**
Just a draft to track progress.
Comments are welcome. No idea if this is a good way of doing such stuff...

I still have the error:
```
xdp-dispatcher.c:5:10: fatal error: 'bpf/bpf_helpers.h' file not found
#include <bpf/bpf_helpers.h>
         ^~~~~~~~~~~~~~~~~~~
1 error generated.
```

### CODE REVIEW

**Comment by PolynomialDivision** [comment]
Ah, I think I have to checkout git submodules. ;)

**Comment by PolynomialDivision** [comment]
Maybe u want to cherry-pick 60c4db670cda9ce4d6b55e2b5454c06db36987c6

**Comment by tohojo** [comment]
For reallocarray() I think it's simpler to just open-code the overflow check and always switch to realloc(), since we're only using it in that one place...

As for the secure_getenv() wrapper, instead of checking specifically for musl, could you please just add a check to 'configure' for that specific function and define it based on that? See check_perf_consume() for an example.

File: configure [hunk 1]
```diff
@@ -242,6 +242,29 @@ EOF
...
+check_secure_getenv()
+{
+    cat >$TMPDIR/secure_getenv.c <<EOF
+#define _GNU_SOURCE
+#include <stdlib.h>
+#define XDP_OBJECT_ENVVAR "LIBXDP_OBJECT_PATH"
+int main(int argc, char **argv) {
+    char *path, **p;
+
+	path = secure_getenv(XDP_OBJECT_ENVVAR);
```
Comment (tohojo) [inline configure:254]: Don't really need the `#define XDP_OBJECT_ENVVAR` or the variables; you can just do `secure_getenv("test");` and not assign the return value anywhere...

File: configure [hunk 2]
```diff
@@ -242,6 +242,29 @@ EOF
...
+    return 0;
+}
+EOF
+    secure_getenv_err=$($CC -o $TMPDIR/secure_getenv $TMPDIR/secure_getenv.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS 2>&1)
```
Comment (tohojo) [inline configure:258]: No need for the LIBBPF_{CFLAGS,LDLIBS}

File: lib/libxdp/libxdp.c [hunk 3]
```diff
@@ -28,6 +28,10 @@
 #include <xdp/libxdp.h>
 #include <xdp/prog_dispatcher.h>
 
+#if !HAVE_SECURE_GETENV
```
Comment (tohojo) [inline lib/libxdp/libxdp.c:31]: `#ifndef` - also you need to add the -D to defines.mk

File: lib/util/secure_getenv.c [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#if !HAVE_SECURE_GETENV
+#include <stdlib.h>
+
+// Source: https://www.openwall.com/lists/musl/2019/05/28/3
+char *secure_getenv(const char *name)
```
Comment (tohojo) [inline lib/util/secure_getenv.c:5]: Just turn this into a 'static inline' function and keep it entirely in the header file?

**Review by tohojo — CHANGES_REQUESTED** [review]
On the right track :)

File: lib/util/secure_getenv.c (hunk 4 above)
Comment (PolynomialDivision) [reply lib/util/secure_getenv.c:5]: Yep. Much better. :)

**Comment by PolynomialDivision** [comment]
> For reallocarray() I think it's simpler to just open-code the overflow check and always switch to realloc(), since we're only using it in that one place...

How would u do the overflow check? I just now switched to realloc Just checking the size after the multiplication? I will try implement something.

**Comment by PolynomialDivision** [comment]
@tohojo Like that? https://github.com/xdp-project/xdp-tools/pull/75/files#diff-4ef1b5443fb0e5b06a85d68a016050cb03d7b40d2d91a08c8d94341d3743c7baR56-R68

File: lib/util/secure_getenv.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#ifndef HAVE_SECURE_GETENV
```
Comment (tohojo) [inline lib/util/secure_getenv.h:1]: Put this around the function definition only - makes it easy to add more compat stuff later if needed...

File: lib/util/params.c [hunk 6]
```diff
@@ -53,9 +53,20 @@ static int handle_multistring(char *optarg, void *tgt, void *typearg)
...
+	num_strings = opt_set->num_strings + 1;
+	max_strings = SIZE_MAX / sizeof(*opt_set->strings);
+	if (max_strings < num_strings) {
+		// overflow
+		ptr = NULL;
+	}
+	else {
+		// no overflow
+		ptr = realloc(opt_set->strings, num_strings);
+	}
```
[the remark this answers is not in the document]
Comment (PolynomialDivision) [reply lib/util/params.c:68]: I guess you mean `	ptr = realloc(opt_set->strings, sizeof(*opt_set->strings) * (opt_set->num_strings +1));` ;)

File: lib/util/secure_getenv.h (hunk 5 above)
Comment (PolynomialDivision) [reply lib/util/secure_getenv.h]: I wrapped that around the include statement, too. Or is that bad practice?

**Review by tohojo — APPROVED** [review]
LGTM - let's wait for CI and see if it agrees :)

**Pull request merged**